core/src/main/java/org/apache/accumulo/core/iterators/system/SourceSwitchingIterator.java

```diff
@@ -126,14 +142,23 @@ public void next() throws IOException {
 
   private void readNext(boolean initialSeek) throws IOException {
 
+    // we need to check here if we were yielded in case the source was switched out and re-seeked by someone else (minor compaction/InMemoryMap)
+    boolean yielded = (yield.isPresent() && yield.get().hasYielded());
+
     // check of initialSeek second is intentional so that it does not short
     // circuit the call to switchSource
-    boolean seekNeeded = (!onlySwitchAfterRow && switchSource()) || initialSeek;
+    boolean seekNeeded = yielded || (!onlySwitchAfterRow && switchSource()) || initialSeek;
 
     if (seekNeeded)
       if (initialSeek)
         iter.seek(range, columnFamilies, inclusive);
-      else
+      else if (yielded) {
+        Key yieldPosition = yield.get().getPositionAndReset();
+        if (!range.contains(yieldPosition)) {
+          throw new IOException("Underlying iterator yielded to a position outside of its range: " + yieldPosition + " not in " + range);
```

Comment:
How does the client handle this IOException being thrown? I'm hoping that it would propagate as a ThriftApplicationException server-side and an AccumuloException client side.